src/Symfony/Component/Serializer/Normalizer/DataUriNormalizer.php

```diff
@@ -0,0 +1,121 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Serializer\Normalizer;
+
+use Symfony\Component\HttpFoundation\File\File;
+use Symfony\Component\HttpFoundation\File\MimeType\MimeTypeGuesser;
+use Symfony\Component\HttpFoundation\File\MimeType\MimeTypeGuesserInterface;
+use Symfony\Component\Serializer\Exception\UnexpectedValueException;
+
+/**
+ * Normalizes a {@see \SplFileInfo} object to a data URI.
+ * Denormalizes a data URI to a {@see \SplFileObject} object using .
+ *
+ * @author Kévin Dunglas <dunglas@gmail.com>
+ */
+class DataUriNormalizer implements NormalizerInterface, DenormalizerInterface
+{
+    /**
+     * @var MimeTypeGuesserInterface
+     */
+    private $mimeTypeGuesser;
+
+    public function __construct(MimeTypeGuesserInterface $mimeTypeGuesser = null)
+    {
+        if (null === $mimeTypeGuesser && class_exists('Symfony\Component\HttpFoundation\File\MimeType\MimeTypeGuesser')) {
+            $mimeTypeGuesser = MimeTypeGuesser::getInstance();
+        }
+
+        $this->mimeTypeGuesser = $mimeTypeGuesser;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function normalize($object, $format = null, array $context = array())
+    {
+        if ($object instanceof File) {
+            $mimeType = $object->getMimeType();
+        } elseif ($this->mimeTypeGuesser) {
+            $mimeType = $this->mimeTypeGuesser->guess($object->getPathname());
+        } else {
+            $mimeType = 'application/octet-stream';
+        }
+
+        list($typeName) = explode('/', $mimeType, 2);
+
+        if (!$object instanceof \SplFileObject) {
+            $object = $object->openFile();
```

Comment:
I agree thought that this reads weird to me. The entire class suffers a little bit of complexity since we're handling SplFileInfo, SplFileObject and File (even though they are all instances of `SplFileInfo`). It might be better to have 2 (or 3) uri normalizers and use a base class to share code. Using some private functions my add clarity (e.g. `private function getMimeType(\SplFileInfo $object)` and `extractSplFileObject(\SplFileInfo $object)`).

After looking for awhile, I don't think there is a problem - it can just be more clear :)